api/providers/WarehouseProvider.cs

```diff
@@ -0,0 +1,108 @@
+using CargoHub.DTOs;
+
+public class WarehouseProvider : ICRUD<Warehouse>
+{
+    private readonly AppDbContext _db;
+    private readonly AddressProvider _addressProvider;
+    private readonly ContactProvider _contactProvider;
+
+    public WarehouseProvider(AppDbContext db, AddressProvider addressProvider, ContactProvider contactProvider)
+    {
+        _db = db;
+        _addressProvider = addressProvider;
+        _contactProvider = contactProvider;
+    }
+
+    public Warehouse? Create<IDTO>(IDTO newElement)
+    {
+        if (newElement is not WarehouseDTO request) throw new Exception("Request invalid");
+
+        // Check if warehouse already exists
+        if (_db.Warehouses.Any(w => w.Name == request.Name))
+            throw new Exception("Warehouse already exists");
+
+        // Case 1 and 2 and 5: Validate ContactId and AddressId if provided
+        if (request.ContactId != null)
+        {
+            var contact = _db.Contacts.Find(request.ContactId);
+            if (contact == null) throw new Exception("ContactID does not exist");
```

Comment:
Graag de `api/providers/ContactProvider.cs : GetById` method gebruiken. In de `WarehouseProvider` maak je al gebruik van de `_contactProvider`.
```suggestion
            if (_contactProvider.GetById(request.ContactId) == null) throw new Exception("ContactID does not exist");
```